File: lib/communication_functions.py
import socket, json, time, struct
import os

# Windows
if os.name == 'nt':
    import msvcrt

# Posix (Linux, OS X)
else:
    import sys
    import termios
    import atexit
    from select import select


class KBHit:
    def __init__(self):
        '''Creates a KBHit object that you can call to do various keyboard things.
        '''
        if os.name == 'nt':
            pass
        else:
            # Save the terminal settings
            self.fd = sys.stdin.fileno()
            self.new_term = termios.tcgetattr(self.fd)
            self.old_term = termios.tcgetattr(self.fd)

            # New terminal setting unbuffered
            self.new_term[3] = (self.new_term[3] & ~termios.ICANON & ~termios.ECHO)
            termios.tcsetattr(self.fd, termios.TCSAFLUSH, self.new_term)
            # Support normal-terminal reset at exit
            atexit.register(self.set_normal_term)
    def set_normal_term(self):
        ''' Resets to normal terminal.  On Windows this is a no-op.
        '''
        if os.name == 'nt':
            pass
        else:
            termios.tcsetattr(self.fd, termios.TCSAFLUSH, self.old_term)

    def getch(self):
        ''' Returns a keyboard character after kbhit() has been called.
            Should not be called in the same program as getarrow().
        '''
        s = ''
        if os.name == 'nt':
            return msvcrt.getch().decode('utf-8')
        else:
            return sys.stdin.read(1)

    def getarrow(self):
        ''' Returns an arrow-key code after kbhit() has been called. Codes are
        0 : up
        1 : right
        2 : down
        3 : left
        Should not be called in the same program as getch().
        '''
        if os.name == 'nt':
            msvcrt.getch() # skip 0xE0
            c = msvcrt.getch()
            vals = [72, 77, 80, 75]
        else:
            c = sys.stdin.read(3)[2]
            vals = [65, 67, 66, 68]
        return vals.index(ord(c.decode('utf-8')))

    def kbhit(self):
        ''' Returns True if keyboard character was hit, False otherwise.
        '''
        if os.name == 'nt':
            return msvcrt.kbhit()

        else:
            dr,dw,de = select([sys.stdin], [], [], 0)
            return dr != []
    def init_term(self):
        self.fd = sys.stdin.fileno()
        self.new_term = termios.tcgetattr(self.fd)
        self.old_term = termios.tcgetattr(self.fd)
        # New terminal setting unbuffered
        self.new_term[3] = (self.new_term[3] & ~termios.ICANON & ~termios.ECHO)
        termios.tcsetattr(self.fd, termios.TCSAFLUSH, self.new_term)
        # Support normal-terminal reset at exit
        atexit.register(self.set_normal_term)

def get_socket():
    socket_path = '/tmp/matrix_connector'
    client = socket.socket(socket.AF_UNIX, socket.SOCK_STREAM)
    client.connect(socket_path)
    return client

def send_message(client, strr):
    message = strr.encode("utf-8")
    try:
        # print("Sending message of length: " + str(len(message)))
        # print("Sending message length...")
        client.sendall(struct.pack("!I", len(message)))
        # print("Sending message...")
        client.sendall(message)
    except:
        # print("Error sending message.")
        exit(1)
    response = client.recv(1024)
    response = json.loads(response)
    if(response["success"]):
        # print("Image successfully sent!!!! :partying-face:")
        return True
    else:
        raise Exception("Something bad happened in the matrix connector")
        return False

File: cpp_lib/communication_functions.cpp
#include "communication_functions.h"
#include <iostream>
#include <cstring>
#include <unistd.h>
#include <json/json.h>

// Function definitions
int get_socket() {
    const char* socket_path = "/tmp/matrix_connector";
    int client = socket(AF_UNIX, SOCK_STREAM, 0);
    if (client < 0) {
        std::cerr << "Error creating socket" << std::endl;
        exit(1);
    }

    struct sockaddr_un addr;
    memset(&addr, 0, sizeof(addr));
    addr.sun_family = AF_UNIX;
    strncpy(addr.sun_path, socket_path, sizeof(addr.sun_path) - 1);

    if (connect(client, (struct sockaddr*)&addr, sizeof(addr)) < 0) {
        std::cerr << "Error connecting to socket" << std::endl;
        close(client);
        exit(1);
    }

    return client;
}

bool send_message(int client, const std::string& strr) {
    std::string message = strr;
    try {
        std::cout << "Sending message of length: " << message.length() << std::endl;
        std::cout << "Sending message length..." << std::endl;
        uint32_t len = htonl(message.length());
        if (send(client, &len, sizeof(len), 0) < 0) {
            throw std::runtime_error("Error sending message length");
        }

        std::cout << "Sending message..." << std::endl;
        if (send(client, message.c_str(), message.length(), 0) < 0) {
            throw std::runtime_error("Error sending message");
        }
    } catch (const std::exception& e) {
        std::cerr << e.what() << std::endl;
        close(client);
        exit(1);
    }

    char buffer[1024];
    ssize_t bytes_received = recv(client, buffer, sizeof(buffer) - 1, 0);
    if (bytes_received < 0) {
        std::cerr << "Error receiving response" << std::endl;
        close(client);
        exit(1);
    }

    buffer[bytes_received] = '\0';
    Json::Value response;
    Json::CharReaderBuilder reader;
    std::string errs;
    std::istringstream s(buffer);
    if (!Json::parseFromStream(reader, s, &response, &errs)) {
        std::cerr << "Error parsing JSON response: " << errs << std::endl;
        close(client);
        exit(1);
    }

    if (response["success"].asBool()) {
        return true;
    } else {
        throw std::runtime_error("Something bad happened in the matrix connector");
        return false;
    }
}